### ISSUE TITLE
Add a check for a perfect initial rewiring before applying rewiring techniques

### ISSUE THREAD

**Issue by ecpeterson** [rigetti/quilc#94]
Right now, quilc will make its best guess for a good rewiring, even when the identity mapping requires no swaps at all. This can sometimes cause unnecessary swaps.

**Comment by notmgsk** [comment]
@ecpeterson do you have an example of when this happens?

**Comment by ecpeterson** [comment]
Yes: have a look at `benchmarking/quil-available/0020q-0000143-3f7b88e4-sampling_quinoa.quil` on Acorn (a/k/a `(build-skew-rectangular-chip 0 4 5)`). If you run it with `PRAGMA INITIAL_REWIRING "NAIVE"` prefixed, it'll compile with no SWAPs. If you run it with the supposedly more intelligent `PRAGMA INITIAL_REWIRING "PARTIAL"`, it'll compile with plenty of SWAPs.

**Comment by notmgsk** [comment]
I'm not confident how this could be solved. My initial thought was to test that the adjacent qubits to each used logical qubit were in the set of used logical qubits: if yes, use the identity rewiring. But that's not quite right.

Hints @ecpeterson?

**Comment by ecpeterson** [comment]
There's more than one way to skin this cat, but I was imagining a check in `prog-initial-rewiring` along the following lines:
```
(when (loop :for instr :across (parsed-program-executable-code parsed-prog)
                       :always (or (not (typep instr 'gate-application))
                                   (lookup-hardware-object-by-qubits chip-spec (application-arguments instr))))
  ;; install the NAIVE rewirer, which will effect no SWAPs
  ;; break early)
;; otherwise, continue making decisions
```
There might be some fuss in that disjunction about dead qubits.

**Comment by ecpeterson** [comment]
I think this is closed by #380.
